core/kernel/transfer_list.c

```diff
@@ -0,0 +1,522 @@
+	dcache_cleaninv_range(tl, sz);
+	unmap_list(tl, sz);
+}
+
+void transfer_list_unmap_nosync(struct transfer_list_header *tl)
+{
+	unmap_list(tl, tl->max_size);
+}
+
+void transfer_list_dump(struct transfer_list_header *tl)
+{
+	struct transfer_list_entry *te = NULL;
+	int i = 0;
+
+	if (!tl)
+		return;
+
+	DMSG("Dump transfer list:");
+	DMSG("signature  %#"PRIx32, tl->signature);
+	DMSG("checksum   %#"PRIx8, tl->checksum);
+	DMSG("version    %#"PRIx8, tl->version);
+	DMSG("hdr_size   %#"PRIx8, tl->hdr_size);
+	DMSG("alignment  %#"PRIx8, tl->alignment);
+	DMSG("size       %#"PRIx32, tl->size);
+	DMSG("max_size   %#"PRIx32, tl->max_size);
+	while (true) {
+		te = transfer_list_next(tl, te);
+		if (!te)
+			break;
+
+		DMSG("Entry %d:", i++);
+		DMSG("tag_id     %#"PRIx16, te->tag_id);
+		DMSG("hdr_size   %#"PRIx8, te->hdr_size);
+		DMSG("data_size  %#"PRIx32, te->data_size);
+		DMSG("data_addr  %#"PRIxVA,
+		     (unsigned long)transfer_list_entry_data(te));
+	}
+}
+
+/*******************************************************************************
+ * Creating a transfer list in a reserved memory region specified
+ * Compliant to 2.4.5 of Firmware handoff specification (v0.9)
+ * Return pointer to the created transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(paddr_t pa, size_t max_size)
+{
+	struct transfer_list_header *tl = NULL;
+
+	if (!pa || max_size == 0)
+		return NULL;
+
+	if (!IS_ALIGNED(pa, TL_ALIGNMENT(TRANSFER_LIST_INIT_MAX_ALIGN)) ||
+	    !IS_ALIGNED(max_size, TL_ALIGNMENT(TRANSFER_LIST_INIT_MAX_ALIGN)) ||
+	    max_size < sizeof(*tl)) {
+		return NULL;
+	}
+
+	tl = core_mmu_add_mapping(MEM_AREA_TRANSFER_LIST, pa, max_size);
+	if (!tl)
+		return NULL;
+
+	memset(tl, 0, max_size);
+	tl->signature = TRANSFER_LIST_SIGNATURE;
+	tl->version = TRANSFER_LIST_VERSION;
+	tl->hdr_size = sizeof(*tl);
+	tl->alignment = TRANSFER_LIST_INIT_MAX_ALIGN; /* initial max align */
+	tl->size = sizeof(*tl); /* initial size is the size of header */
+	tl->max_size = max_size;
+
+	transfer_list_update_checksum(tl);
+
+	return tl;
+}
+
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return pointer to the relocated transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *
+transfer_list_relocate(struct transfer_list_header *tl, paddr_t pa,
+		       size_t max_size)
+{
+	paddr_t new_addr = 0;
+	struct transfer_list_header *new_tl = NULL;
+	size_t new_max_size = 0;
+
+	if (!tl || !pa || max_size == 0)
+		return NULL;
+
+	new_addr = TL_ADDR_ALIGN(pa, virt_to_phys(tl),
+				 TL_ALIGNMENT(tl->alignment));
+	new_max_size = max_size - (new_addr - pa);
+
+	/* the new space is not sufficient for the tl */
```

Comment:
Nitpicking: prefer `/* The new space is not sufficient for the TL */`

Reply:
Done.